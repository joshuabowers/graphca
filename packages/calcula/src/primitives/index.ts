import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
