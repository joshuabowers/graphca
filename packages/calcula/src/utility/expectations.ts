import { method, multi, Multi, _ } from '@arrows/multimethod'
import { isWriter, Writer } from '../monads/writer'
import { TreeNode } from './tree'
import { Real, Complex, isReal, isComplex } from '../primitives'

export type ExpectCloseTo = Multi 
  & ((actual: Writer<Real>, expected: Writer<Real>, precision: number) => void)
  & ((actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => void)

export const expectCloseTo: ExpectCloseTo = multi(
  method(
    [isReal, isReal, _], 
    (actual: Writer<Real>, expected: Writer<Real>, precision: number) => 
      expect(actual.value.value).toBeCloseTo(expected.value.value, precision)
  ),
  method(
    [isComplex, isComplex, _],
    (actual: Writer<Complex>, expected: Writer<Complex>, precision: number) => {
      expect(actual.value.a).toBeCloseTo(expected.value.a, precision)
      expect(actual.value.b).toBeCloseTo(expected.value.b, precision)
    }
  )
)

type Input = TreeNode|Writer<TreeNode>|[Input, Input]
export type Operation = [Input, string]

export const expectWriter = <
  Actual extends TreeNode, 
  Expected extends TreeNode|Writer<TreeNode>
>(
  actual: Writer<Actual>
) => (expected: Expected, ...operations: Operation[]) => {
  expect(actual).toEqual({
    value: isWriter(expected) ? expected.value : expected,
    log: operations.map(([input, action]) => ({
      input: isWriter(input) 
        ? input.value 
        : (Array.isArray(input) 
          ? input.map(i => isWriter(i) ? i.value : i)
          : input), 
      action
    }))
  })
}
