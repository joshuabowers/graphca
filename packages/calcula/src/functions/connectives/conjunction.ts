import { _ } from '@arrows/multimethod'
import { Writer, unit } from "../../monads/writer"
import { Clades, Genera, Species, isSpecies, isGenus } from "../../utility/tree"
import { Real, Complex, Boolean, boolean } from "../../primitives"
import { BinaryNode, binary, when, binaryInfixRule } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { isComplement } from './complement'
import { isDisjunction } from './disjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Conjunction = Connective<Species.and>

export const andRule = binaryInfixRule(Unicode.and)

export const [and, isConjunction, $and] = binary<Conjunction, Boolean>(
  Species.and, Genera.connective
)(
  (l, r) => [boolean(l.value !== 0 && r.value !== 0), andRule(l, r), 'real conjunction'],
  (l, r) => [
    boolean((l.a !== 0 || l.b !== 0) && (r.a !== 0 || r.b !== 0)), 
    andRule(l, r),
    'complex conjunction'
  ],
  (l, r) => [boolean(l.value && r.value), andRule(l, r), 'boolean conjunction']
)(
  when(
    [_, isValue(boolean(true))], 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive identity']
  ),
  when(
    [isValue(boolean(true)), _], 
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive identity']
  ),
  when(
    [_, isValue(boolean(false))], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    [isValue(boolean(false)), _], 
    [boolean(false), rule`${boolean(false)}`, 'conjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive idempotency']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [unit(l), rule`${l}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isDisjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [unit(r), rule`${r}`, 'conjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(false), rule`${boolean(false)}`, 'contradiction']
  )
)
