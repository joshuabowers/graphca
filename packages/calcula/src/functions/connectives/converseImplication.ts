import { _ } from '@arrows/multimethod'
import { unit } from "../../monads/writer"
import { Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when, binaryInfixRule } from "../../closures/binary"
import { isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { not, notRule } from './complement'
import { or } from './disjunction'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type ConverseImplication = Connective<Species.converse>

export const converseRule = binaryInfixRule(Unicode.converse)

export const [converse, isConverseImplication, $converse] = binary<ConverseImplication, Boolean>(
  Unicode.converse, Notation.infix, Species.converse, Genera.connective
)(
  (l, r) => or(unit(l), not(unit(r))),
  (l, r) => or(unit(l), not(unit(r))),
  (l, r) => or(unit(l), not(unit(r)))
)(
  when(
    [isValue(boolean(true)), _], 
    [boolean(true), rule`${boolean(true)}`, 'converse implication annihilator']
  ),
  when(
    [_, isValue(boolean(true))], 
    (l, _r) => [unit(l), rule`${l}`, 'converse implication identity']
  ),
  when(
    [isValue(boolean(false)), _], 
    (_l, r) => [not(unit(r)), notRule(r), 'converse implication complementation']
  ),
  when(
    [_, isValue(boolean(false))], 
    [boolean(true), rule`${boolean(true)}`, 'converse implication annihilator']
  )
)
