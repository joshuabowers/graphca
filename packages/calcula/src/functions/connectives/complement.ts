import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation, isSpecies } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { Unary, unary, unaryFnRule, when } from "../../closures/unary"
import { Conjunction, and, andRule } from './conjunction'
import { Disjunction, or, orRule } from './disjunction'
import { ExclusiveDisjunction, xor, xorRule } from './exclusiveDisjunction'
import { Implication } from './implication'
import { AlternativeDenial, nand, nandRule } from './alternativeDenial'
import { JointDenial, nor, norRule } from './jointDenial'
import { Biconditional, xnor, xnorRule } from './biconditional'
import { ConverseImplication } from './converseImplication'
import { Unicode } from '../../Unicode'
import { rule } from '../../utility/rule'

export type Complement = Unary<Species.not, Genera.connective>

export const notRule = unaryFnRule(Unicode.not)

export const [not, isComplement, $not] = unary<Complement, Boolean>(
  Unicode.not, Notation.prefix, Species.not, Genera.connective
)(
  r => [boolean(r.value === 0), notRule(r), 'real complement'],
  c => [boolean(c.a === 0 && c.b === 0), notRule(c), 'complex complement'],
  b => [boolean(!b.value), notRule(b), 'boolean complement']
)(
  // NOTE: Cannot use derived isSpecies guards as they are not defined
  // by this point.
  when<Complement>(
    isSpecies(Species.not), 
    v => [v.expression, rule`${v.expression}`, 'double complement']
  ),
  when<Conjunction>(
    isSpecies(Species.and), 
    v => [nand(v.left, v.right), nandRule(v.left, v.right), 'complement of conjunction']
  ),
  when<Disjunction>(
    isSpecies(Species.or), 
    v => [nor(v.left, v.right), norRule(v.left, v.right), 'complement of disjunction']
  ),
  when<AlternativeDenial>(
    isSpecies(Species.nand), 
    v => [and(v.left, v.right), andRule(v.left, v.right), 'complement of alternative denial']
  ),
  when<JointDenial>(
    isSpecies(Species.nor), 
    v => [or(v.left, v.right), orRule(v.left, v.right), 'complement of joint denial']
  ),
  when<ExclusiveDisjunction>(
    isSpecies(Species.xor), 
    v => [xnor(v.left, v.right), xnorRule(v.left, v.right), 'complement of exclusive disjunction']
  ),
  when<Implication>(
    isSpecies(Species.implies),
    v => [
      and(v.left, not(v.right)), 
      rule`${v.left} ${Unicode.and} (${v.right})${Unicode.not}`, 
      'complement of implication'
    ]
  ),
  when<Biconditional>(
    isSpecies(Species.xnor),
    v => [xor(v.left, v.right), xorRule(v.left, v.right), 'complement of biconditional']
  ),
  when<ConverseImplication>(
    isSpecies(Species.converse),
    v => [
      and(not(v.left), v.right), 
      rule`(${v.left})${Unicode.not} ${Unicode.and} ${v.right}`,
      'complement of converse implication'
    ]
  )
)
