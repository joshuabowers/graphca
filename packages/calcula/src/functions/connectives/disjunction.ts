import { _ } from '@arrows/multimethod'
import { TreeNode, Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when } from "../../closures/binary"
import { deepEquals, isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { Complement, isComplement } from './complement'
import { isConjunction } from './conjunction'
import { implies } from './implication'
import { converse } from './converseImplication'
import { Unicode } from '../../Unicode'

export type Disjunction = Connective<Species.or>

export const [or, isDisjunction, $or] = binary<Disjunction, Boolean>(
  Unicode.or, Notation.infix, Species.or, Genera.connective
)(
  (l, r) => boolean(l.value.value !== 0 || r.value.value !== 0),
  (l, r) => boolean(
    l.value.a !== 0 || l.value.b !== 0 || r.value.a !== 0 || r.value.b !== 0
  ),
  (l, r) => boolean(l.value.value || r.value.value)
)(
  when(
    [_, isValue(boolean(false))],
    (l, _r) => [l, 'disjunctive identity']
  ),
  when(
    [isValue(boolean(false)), _],
    (_l, r) => [r, 'disjunctive identity']
  ),
  when(
    [_, isValue(boolean(true))],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    [isValue(boolean(true)), _],
    [boolean(true), 'disjunctive annihilator']
  ),
  when(
    deepEquals, 
    (l, _r) => [l, 'disjunctive idempotency']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.left),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(r) && deepEquals(l, r.value.right),
    (l, _r) => [l, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.left, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isConjunction(l) && deepEquals(l.value.right, r),
    (_l, r) => [r, 'disjunctive absorption']
  ),
  when(
    (l, r) => isComplement(r) && deepEquals(l, r.value.expression),
    [boolean(true), 'disjunctive tautology']
  ),
  when(
    (l, r) => isComplement(l) && deepEquals(l.value.expression, r),
    [boolean(true), 'disjunctive tautology']
  ),
  when<Complement, TreeNode>(
    (l, _r) => isComplement(l),
    (l, r) => [
      implies(l.value.expression, r), 
      'disjunctive implication'
    ]
  ),
  when<TreeNode, Complement>(
    (_l, r) => isComplement(r),
    (l, r) => [
      converse(l, r.value.expression), 
      'disjunctive converse implication'
    ]
  )
)
