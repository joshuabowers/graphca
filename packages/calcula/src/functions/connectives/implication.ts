import { _ } from '@arrows/multimethod'
import { Genera, Species, Notation } from "../../utility/tree"
import { Boolean, boolean } from "../../primitives"
import { binary, when } from "../../closures/binary"
import { isValue } from "../../utility/deepEquals"
import { Connective } from './connective'
import { not } from './complement'
import { or } from './disjunction'
import { Unicode } from '../../Unicode'

export type Implication = Connective<Species.implies>

export const [implies, isImplication, $implies] = binary<Implication, Boolean>(
  Unicode.implies, Notation.infix, Species.implies, Genera.connective
)(
  (l, r) => or(not(l), r),
  (l, r) => or(not(l), r),
  (l, r) => or(not(l), r)
)(
  when(
    [isValue(boolean(true)), _], 
    (_l, r) => [r, 'implicative identity']
  ),
  when(
    [_, isValue(boolean(true))], 
    [boolean(true), 'implicative annihilator']
  ),
  when(
    [isValue(boolean(false)), _], 
    [boolean(true), 'implicative annihilator']
  ),
  when(
    [_, isValue(boolean(false))], 
    (l, _r) => [
      not(l), 'implicative complementation'
    ]
  )
)
