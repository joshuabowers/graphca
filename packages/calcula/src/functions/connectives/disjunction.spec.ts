import { unit } from '../../monads/writer'
import { expectToEqualWithSnapshot } from '../../utility/expectations'
import { Clades, Genera, Species } from '../../utility/tree'
import { real, complex, boolean } from '../../primitives'
import { variable } from '../../variable'
import { not } from './complement'
import { and } from './conjunction'
import { implies } from './implication'
import { converse } from './converseImplication'
import { or, $or } from './disjunction'

describe('$or', () => {
  it('generates a Disjunction for a pair of TreeNode inputs', () => {
    expect(
      $or(unit(variable('x').value), unit(variable('y').value))[0]
    ).toEqual({
      clade: Clades.binary, genus: Genera.connective, species: Species.or,
      left: unit(variable('x').value), right: unit(variable('y').value)
    })
  })
})

describe('or', () => {
  it('returns true when given two true things', () => {
    expectToEqualWithSnapshot(
      or(boolean(true), boolean(true)),
      boolean(true)
    )
  })

  it('returns true if the left argument is true', () => {
    expectToEqualWithSnapshot(
      or(boolean(true), boolean(false)),
      boolean(true)
    )
  })

  it('returns true if the right argument is true', () => {
    expectToEqualWithSnapshot(
      or(boolean(false), boolean(true)),
      boolean(true)
    )
  })

  it('returns false if both arguments are false', () => {
    expectToEqualWithSnapshot(
      or(boolean(false), boolean(false)),
      boolean(false)
    )
  })

  it('casts reals to booleans, where 0 is false, non-zero is true', () => {
    expectToEqualWithSnapshot(
      or(real(5), real(0)),
      boolean(true)
    )
  })

  it('casts complexes to booleans, 0 => false, non-zero => true', () => {
    expectToEqualWithSnapshot(
      or(complex([5, 0]), complex([0, 0])),
      boolean(true)
    )
  })

  it('returns the left operand if the right is false', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), boolean(false)),
      variable('x')
    )
  })

  it('returns the right operand if the left is false', () => {
    expectToEqualWithSnapshot(
      or(boolean(false), variable('x')),
      variable('x')
    )
  })

  it('returns true if the right operand is true', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), boolean(true)),
      boolean(true)
    )
  })

  it('returns true if the left operand is true', () => {
    expectToEqualWithSnapshot(
      or(boolean(true), variable('x')),
      boolean(true)
    )
  })

  it('returns the left operand if left is equivalent to right', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), variable('x')),
      variable('x')
    )
  })

  it('returns the left operand if the right is a left-child Conjunction of the left', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), and(variable('x'), variable('y'))),
      variable('x')
    )
  })

  it('returns the left operand if the right is a right-child Conjunction of the left', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), and(variable('y'), variable('x'))),
      variable('x')
    )
  })

  it('returns the right operand if the left is a left-child Conjunction of the right', () => {
    expectToEqualWithSnapshot(
      or(and(variable('x'), variable('y')), variable('x')),
      variable('x')
    )
  })

  it('returns the right operand if the left is a right-child Conjunction of the right', () => {
    expectToEqualWithSnapshot(
      or(and(variable('y'), variable('x')), variable('x')),
      variable('x')
    )
  })

  it('returns true if the right operand is the negation of the left', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), not(variable('x'))),
      boolean(true)
    )
  })

  it('returns true if the left operand is the negation of the right', () => {
    expectToEqualWithSnapshot(
      or(not(variable('x')), variable('x')),
      boolean(true)
    )
  })

  it('returns an implication if the left operand is a complement', () => {
    expectToEqualWithSnapshot(
      or(not(variable('x')), variable('y')),
      implies(variable('x'), variable('y'))
    )
  })

  it('returns a converse if the right operand is a complement', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), not(variable('y'))),
      converse(variable('x'), variable('y'))
    )
  })

  it('returns a Disjunction on variable input', () => {
    expectToEqualWithSnapshot(
      or(variable('x'), variable('y')),
      $or(variable('x'), variable('y'))[0]
    )
  })
})
