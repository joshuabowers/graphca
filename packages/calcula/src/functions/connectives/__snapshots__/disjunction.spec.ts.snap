// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`or casts complexes to booleans, 0 => false, non-zero => true 1`] = `
[
  [
    "(5+0𝒊∨0+0𝒊)",
    "identified disjunction",
  ],
  [
    "([5+0𝒊]∨0+0𝒊)",
    "processing left operand",
  ],
  [
    "5+0𝒊",
    "created complex",
  ],
  [
    "({5+0𝒊}∨[0+0𝒊])",
    "processed left operand; processing right operand",
  ],
  [
    "0+0𝒊",
    "created complex",
  ],
  [
    "(5+0𝒊∨{0+0𝒊})",
    "processed right operand",
  ],
  [
    "(5+0𝒊∨0+0𝒊)",
    "complex disjunction",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or casts reals to booleans, where 0 is false, non-zero is true 1`] = `
[
  [
    "(5∨0)",
    "identified disjunction",
  ],
  [
    "([5]∨0)",
    "processing left operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}∨[0])",
    "processed left operand; processing right operand",
  ],
  [
    "0",
    "created real",
  ],
  [
    "(5∨{0})",
    "processed right operand",
  ],
  [
    "(5∨0)",
    "real disjunction",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns a Disjunction on variable input 1`] = `
[
  [
    "(x∨y)",
    "identified disjunction",
  ],
  [
    "([x]∨y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∨{y})",
    "processed right operand",
  ],
  [
    "(x∨y)",
    "created disjunction",
  ],
]
`;

exports[`or returns a converse if the right operand is a complement 1`] = `
[
  [
    "(x∨¬(y))",
    "identified disjunction",
  ],
  [
    "([x]∨¬(y))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[¬(y)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(y)",
    "identified complement",
  ],
  [
    "¬([y])",
    "processing argument",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "¬({y})",
    "processed argument",
  ],
  [
    "¬(y)",
    "created complement",
  ],
  [
    "(x∨{¬(y)})",
    "processed right operand",
  ],
  [
    "(x∨¬(y))",
    "disjunctive converse implication",
  ],
  [
    "(x←y)",
    "identified converse implication",
  ],
  [
    "([x]←y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}←[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x←{y})",
    "processed right operand",
  ],
  [
    "(x←y)",
    "created converse implication",
  ],
]
`;

exports[`or returns an implication if the left operand is a complement 1`] = `
[
  [
    "(¬(x)∨y)",
    "identified disjunction",
  ],
  [
    "([¬(x)]∨y)",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}∨[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(¬(x)∨{y})",
    "processed right operand",
  ],
  [
    "(¬(x)∨y)",
    "disjunctive implication",
  ],
  [
    "(x→y)",
    "identified implication",
  ],
  [
    "([x]→y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}→[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x→{y})",
    "processed right operand",
  ],
  [
    "(x→y)",
    "created implication",
  ],
]
`;

exports[`or returns false if both arguments are false 1`] = `
[
  [
    "(false∨false)",
    "identified disjunction",
  ],
  [
    "([false]∨false)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}∨[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(false∨{false})",
    "processed right operand",
  ],
  [
    "(false∨false)",
    "disjunctive identity",
  ],
  [
    "false",
    "created boolean",
  ],
]
`;

exports[`or returns the left operand if left is equivalent to right 1`] = `
[
  [
    "(x∨x)",
    "identified disjunction",
  ],
  [
    "([x]∨x)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(x∨{x})",
    "processed right operand",
  ],
  [
    "(x∨x)",
    "disjunctive idempotency",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns the left operand if the right is a left-child Conjunction of the left 1`] = `
[
  [
    "(x∨(x∧y))",
    "identified disjunction",
  ],
  [
    "([x]∨(x∧y))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[(x∧y)])",
    "processed left operand; processing right operand",
  ],
  [
    "(x∧y)",
    "identified conjunction",
  ],
  [
    "([x]∧y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∧{y})",
    "processed right operand",
  ],
  [
    "(x∧y)",
    "created conjunction",
  ],
  [
    "(x∨{(x∧y)})",
    "processed right operand",
  ],
  [
    "(x∨(x∧y))",
    "disjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns the left operand if the right is a right-child Conjunction of the left 1`] = `
[
  [
    "(x∨(y∧x))",
    "identified disjunction",
  ],
  [
    "([x]∨(y∧x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[(y∧x)])",
    "processed left operand; processing right operand",
  ],
  [
    "(y∧x)",
    "identified conjunction",
  ],
  [
    "([y]∧x)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(y∧{x})",
    "processed right operand",
  ],
  [
    "(y∧x)",
    "created conjunction",
  ],
  [
    "(x∨{(y∧x)})",
    "processed right operand",
  ],
  [
    "(x∨(y∧x))",
    "disjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns the left operand if the right is false 1`] = `
[
  [
    "(x∨false)",
    "identified disjunction",
  ],
  [
    "([x]∨false)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(x∨{false})",
    "processed right operand",
  ],
  [
    "(x∨false)",
    "disjunctive identity",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns the right operand if the left is a left-child Conjunction of the right 1`] = `
[
  [
    "((x∧y)∨x)",
    "identified disjunction",
  ],
  [
    "([(x∧y)]∨x)",
    "processing left operand",
  ],
  [
    "(x∧y)",
    "identified conjunction",
  ],
  [
    "([x]∧y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∧[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x∧{y})",
    "processed right operand",
  ],
  [
    "(x∧y)",
    "created conjunction",
  ],
  [
    "({(x∧y)}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "((x∧y)∨{x})",
    "processed right operand",
  ],
  [
    "((x∧y)∨x)",
    "disjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns the right operand if the left is a right-child Conjunction of the right 1`] = `
[
  [
    "((y∧x)∨x)",
    "identified disjunction",
  ],
  [
    "([(y∧x)]∨x)",
    "processing left operand",
  ],
  [
    "(y∧x)",
    "identified conjunction",
  ],
  [
    "([y]∧x)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}∧[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(y∧{x})",
    "processed right operand",
  ],
  [
    "(y∧x)",
    "created conjunction",
  ],
  [
    "({(y∧x)}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "((y∧x)∨{x})",
    "processed right operand",
  ],
  [
    "((y∧x)∨x)",
    "disjunctive absorption",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns the right operand if the left is false 1`] = `
[
  [
    "(false∨x)",
    "identified disjunction",
  ],
  [
    "([false]∨x)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(false∨{x})",
    "processed right operand",
  ],
  [
    "(false∨x)",
    "disjunctive identity",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`or returns true if the left argument is true 1`] = `
[
  [
    "(true∨false)",
    "identified disjunction",
  ],
  [
    "([true]∨false)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}∨[false])",
    "processed left operand; processing right operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "(true∨{false})",
    "processed right operand",
  ],
  [
    "(true∨false)",
    "disjunctive identity",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns true if the left operand is the negation of the right 1`] = `
[
  [
    "(¬(x)∨x)",
    "identified disjunction",
  ],
  [
    "([¬(x)]∨x)",
    "processing left operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "({¬(x)}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(¬(x)∨{x})",
    "processed right operand",
  ],
  [
    "(¬(x)∨x)",
    "disjunctive tautology",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns true if the left operand is true 1`] = `
[
  [
    "(true∨x)",
    "identified disjunction",
  ],
  [
    "([true]∨x)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}∨[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(true∨{x})",
    "processed right operand",
  ],
  [
    "(true∨x)",
    "disjunctive annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns true if the right argument is true 1`] = `
[
  [
    "(false∨true)",
    "identified disjunction",
  ],
  [
    "([false]∨true)",
    "processing left operand",
  ],
  [
    "false",
    "created boolean",
  ],
  [
    "({false}∨[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(false∨{true})",
    "processed right operand",
  ],
  [
    "(false∨true)",
    "disjunctive identity",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns true if the right operand is the negation of the left 1`] = `
[
  [
    "(x∨¬(x))",
    "identified disjunction",
  ],
  [
    "([x]∨¬(x))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[¬(x)])",
    "processed left operand; processing right operand",
  ],
  [
    "¬(x)",
    "identified complement",
  ],
  [
    "¬([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "¬({x})",
    "processed argument",
  ],
  [
    "¬(x)",
    "created complement",
  ],
  [
    "(x∨{¬(x)})",
    "processed right operand",
  ],
  [
    "(x∨¬(x))",
    "disjunctive tautology",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns true if the right operand is true 1`] = `
[
  [
    "(x∨true)",
    "identified disjunction",
  ],
  [
    "([x]∨true)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}∨[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(x∨{true})",
    "processed right operand",
  ],
  [
    "(x∨true)",
    "disjunctive annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`or returns true when given two true things 1`] = `
[
  [
    "(true∨true)",
    "identified disjunction",
  ],
  [
    "([true]∨true)",
    "processing left operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true}∨[true])",
    "processed left operand; processing right operand",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "(true∨{true})",
    "processed right operand",
  ],
  [
    "(true∨true)",
    "disjunctive annihilator",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;
